zyte_common_items/util.py

```diff
@@ -54,3 +60,13 @@ def get_origin(tp) -> Tuple:
     typing.get_origin(tp) is only available startingo on 3.8.
     """
     return getattr(tp, "__origin__", ())
+
+
+def is_url_obj(url: Union[str, _Url]) -> TypeGuard[_Url]:
+    return True if isinstance(url, _Url) else False
+
+
+def url_to_str(url: Union[str, _Url]) -> object:
```

Comment:
Type annotations don't look right - it seems this function accepts Any, and returns object of the same type, except for URLs, which are demoted to str. 

What's the use case for returning the url as-is if it's not a string? Why not raise an error?

Reply:
>  it seems this function accepts Any

I'm afraid I don't follow since it's annotated to accept `Union[str, _Url]`.

> What's the use case for returning the url as-is if it's not a string? Why not raise an error?

This setup was intended for static type checks only. Should we also check do some runtime checks with errors during the conversion process?

Reply:
It seems reasonable to fail with a ValueError if conversion to str can't happen.

Reply:
Gotcha, I've updated the logic here: https://github.com/zytedata/zyte-common-items/pull/11/commits/3fee61364f5f6a93598edc0bddbc7618240d6f29

Reply:
I think type annotation now needs to be changed - the output is str, not object.

I'm also not sure about using object in a first place, it seems Any was a better fit.

Reply:
> I think type annotation now needs to be changed - the output is str, not object

Good catch. I've updated it in https://github.com/zytedata/zyte-common-items/pull/11/commits/fbb6ed028b46dfbad6bf8b22747d2f07d087c386.

> I'm also not sure about using object in a first place, it seems Any was a better fit.

Using `Any` would "turn off" the type checker ([ref](https://docs.python.org/3/library/typing.html#the-any-type)): "Notice that no typechecking is performed when assigning a value of type [Any](https://docs.python.org/3/library/typing.html#typing.Any) to a more precise type"

There's also this discussion that I found here: https://stackoverflow.com/questions/39817081/typing-any-vs-object

Reply:
This link suggests 

> That means when the type of a value is object, a type checker will reject almost all operations on it, and assigning it to a variable (or using it as a return value) of a more specialized type is a type error.

it doesn't seem right; likely we could have used a TypeVar to annotate `url`, and make a return type a Union[T, str] -but that's irrelevant now :)

Reply:
Ahhh that's a good point. I missed out on this since the code wasn't accessing any methods from the object. 👍